File: src/main/java/com/clinic/domain/ports/AppointmentEventPublisher.java
package com.clinic.domain.ports;

import com.clinic.domain.entities.Appointment;

/**
 * Port for publishing appointment events to the messaging backbone.
 *
 * <p>AWS project used SNS (topic) -> SQS (queue) for fan-out, and EventBridge for the completion
 * event. Here the adapter targets Azure Service Bus (topic + queues) and Event Grid, but the
 * application layer doesn't know that.
 */
public interface AppointmentEventPublisher {

  /** Publishes the "appointment created" event for country-specific processing. */
  void publishCreated(Appointment appointment);

  /** Publishes the "appointment completed" event (EventBridge -> Event Grid). */
  void publishCompleted(Appointment appointment);

  /** Publishes the "appointment cancelled" event. */
  void publishCancelled(Appointment appointment);
}

File: clinic/domain/ports/appointment_event_publisher.py
"""Port for publishing appointment events to the messaging backbone.

The AWS sibling project used SNS(topic)->SQS(queue) for fan-out, and EventBridge for the
completion event. Here the adapter targets Azure Service Bus, but the application layer
doesn't know that.
"""

from __future__ import annotations

from typing import Protocol, runtime_checkable

from clinic.domain.entities.appointment import Appointment


@runtime_checkable
class AppointmentEventPublisher(Protocol):
    def publish_created(self, appointment: Appointment) -> None:
        """Publishes the "appointment created" event for country-specific processing."""

    def publish_completed(self, appointment: Appointment) -> None:
        """Publishes the "appointment completed" event."""

    def publish_cancelled(self, appointment: Appointment) -> None:
        """Publishes the "appointment cancelled" event."""
